hpke/marshal.go

```diff
@@ -0,0 +1,154 @@
+package hpke
+
+import (
+	"errors"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// marshal serializes an HPKE context.
+func (c *encdecContext) marshal() ([]byte, error) {
+	var b cryptobyte.Builder
+	b.AddUint16(uint16(c.suite.KemID))
+	b.AddUint16(uint16(c.suite.KdfID))
+	b.AddUint16(uint16(c.suite.AeadID))
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.exporterSecret)
+	})
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.key)
+	})
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.baseNonce)
+	})
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.sequenceNumber)
+	})
+	return b.Bytes()
+}
+
+// unmarshalContext parses an HPKE context.
+func unmarshalContext(raw []byte) (*encdecContext, error) {
+	var (
+		err                  error
+		kemID, kdfID, aeadID uint16
+		t                    cryptobyte.String
+	)
+
+	c := new(encdecContext)
+	s := cryptobyte.String(raw)
+	if !s.ReadUint16(&kemID) ||
+		!s.ReadUint16(&kdfID) ||
+		!s.ReadUint16(&aeadID) ||
+		!s.ReadUint8LengthPrefixed(&t) ||
+		!t.ReadBytes(&c.exporterSecret, len(t)) ||
+		!s.ReadUint8LengthPrefixed(&t) ||
+		!t.ReadBytes(&c.key, len(t)) ||
+		!s.ReadUint8LengthPrefixed(&t) ||
+		!t.ReadBytes(&c.baseNonce, len(t)) ||
+		!s.ReadUint8LengthPrefixed(&t) ||
+		!t.ReadBytes(&c.sequenceNumber, len(t)) {
+		return nil, errors.New("failed to parse context")
+	}
+
+	c.suite.KemID = KemID(kemID)
+	c.suite.KdfID = KdfID(kdfID)
+	c.suite.AeadID = AeadID(aeadID)
+	if !c.suite.isValid() {
+		return nil, errors.New("invalid suite")
+	}
+
+	Nh := c.suite.KdfID.Hash().Size()
+	if len(c.exporterSecret) != Nh {
+		return nil, errors.New("invalid exporter secret")
+	}
+
+	Nk := int(c.suite.AeadID.KeySize())
+	if len(c.key) != Nk {
+		return nil, errors.New("invalid key")
+	}
+
+	c.AEAD, err = c.suite.AeadID.New(c.key)
+	if err != nil {
+		return nil, err
+	}
+
+	Nn := c.AEAD.NonceSize()
+	if len(c.baseNonce) != Nn {
+		return nil, errors.New("invalid base nonce")
```

Comment:
invalid nonce length?

Reply:
Changed to "invalid base nonce length". (Made a similar change elsewhere in this function.)